hasktorch/test/ScriptSpec.hs

```diff
@@ -1,32 +1,109 @@
 {-# LANGUAGE NoMonomorphismRestriction #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE DeriveGeneric #-}
 
 module ScriptSpec(spec) where
 
 import Prelude hiding (abs, exp, floor, log, min, max)
 
 import Test.Hspec
-import Control.Exception.Safe
-
-import Torch.Tensor
-import Torch.DType
-import Torch.Layout
-import Torch.TensorFactories
-import Torch.Functional
-import Torch.TensorOptions
+
+import Torch
 import Torch.Script
+import Torch.NN
+import Torch.Autograd
+import GHC.Generics
+
+data MLPSpec = MLPSpec {
+    inputFeatures :: Int,
+    hiddenFeatures0 :: Int,
+    hiddenFeatures1 :: Int,
+    outputFeatures :: Int
+    } deriving (Show, Eq)
+
+data MLP = MLP {
+    l0 :: Linear,
+    l1 :: Linear,
+    l2 :: Linear
+    } deriving (Generic, Show)
+
+instance Parameterized MLP
+instance Randomizable MLPSpec MLP where
+    sample MLPSpec {..} = MLP 
+        <$> sample (LinearSpec inputFeatures hiddenFeatures0)
+        <*> sample (LinearSpec hiddenFeatures0 hiddenFeatures1)
+        <*> sample (LinearSpec hiddenFeatures1 outputFeatures)
+
+mlp :: MLP -> Tensor -> Tensor
+mlp MLP{..} input = 
+    logSoftmax 1
+    . linear l2
+    . relu
+    . linear l1
+    . relu
+    . linear l0
+    $ input
+
+params :: Parameterized a => a -> [Tensor]
+params = (map toDependent).flattenParameters
+
+fromParams :: Parameterized a => a -> [Tensor] -> a
+fromParams init' ps = replaceParameters init' (map IndependentTensor ps)
 
 spec :: Spec
-spec = do
+spec = describe "torchscript" $ do
   it "define and run" $ do
+    let v00 = asTensor (4::Float)
     m <- newModule "m"
+    registerParameter m "p0" v00 False
     define m $
       "def foo(self, x):\n" ++ 
-      "    return (1, 2, x + 3)\n" ++ 
+      "    return (1, 2, x + 3 + self.p0)\n" ++ 
       "\n" ++ 
       "def forward(self, x):\n" ++ 
       "    tuple = self.foo(x)\n" ++ 
       "    return tuple\n"
-    IVTuple [IVInt a,IVInt b,IVTensor c] <- run_method1 m "forward" (IVTensor (ones' []))
+    sm <- toScriptModule m
+    let IVTuple [IVInt a,IVInt b,IVTensor c] = runMethod1 sm "forward" (IVTensor (ones' []))
     a `shouldBe` 1
     b `shouldBe` 2
-    (asValue c :: Float) `shouldBe` 4.0
+    (asValue c :: Float) `shouldBe` 8.0
+    -- save sm "self.pt"
+    -- sm <- load "self.pt"
+    -- let IVTuple [IVInt a,IVInt b,IVTensor c] = runMethod1 sm "forward" (IVTensor (ones' []))
+    -- let [g] =  grad c (flattenParameters sm)
+    -- (asValue g :: Float) `shouldBe` 1.0
+    -- -- This throws 'CppStdException "Exception: Differentiated tensor not require grad; type: std::runtime_error"'.
+    -- -- See libtorch-ffi/src/Torch/Internal/Unmanaged/Autograd.hs
```

Comment:
@apaszke 
This code throws a exception for '!roots[0].function == true' of Autograd.hs.
Do you think it is related to callback's outputs not forced?